helper/compiler_factory.py

```diff
@@ -0,0 +1,92 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+
+import tarfile
+import os
+import re
+import importlib
+from urllib.request import urlretrieve
+from pathlib import Path
+from helper.cd import cd
+from shutil import which
+
+
+class CompilerFactory(object):
+    def __init__(self, toolchain_url, toolchain_extractpath):
+        self.toolchain_url = toolchain_url
+        self.toolchain_extractpath = toolchain_extractpath
+
+    def getCompiler(self):
+        if self.toolchain_url.find('http') != -1 or self.toolchain_url.find('ftp') != -1:
+            extracted_tar = self._downloadToolchain()
+            return self._fetchCompiler(extracted_tar)
+        elif self.toolchain_url.find('g++') != -1:
+            return self._fetch_system('g++')
+        elif self.toolchain_url.find('gcc') != -1:
+            return self._fetch_system('gcc')
+        elif self.toolchain_url.find('clang') != -1:
+            return self._fetch_system('clang')
+        elif self.toolchain_url.find('clang++') != -1:
+            return self._fetch_system('clang++')
+
+    def _fetch_system(self, compiler):
+        compiler_path = which(compiler)
+        if compiler_path is not None:
+            return self._getCompilerFromBinaries(compiler_path)
+        else:
+            raise ImportError('Compiler %s not installed'% compiler)
+
+    def _downloadToolchain(self):
+        with cd(self.toolchain_extractpath):
+            filename, headers = urlretrieve(self.toolchain_url)
+            before = os.listdir()
+            tarball = tarfile.open(filename)
+            tarball.extractall()
+            after = os.listdir()
+            filename = [x for x in after if x not in before]
+            print(self.toolchain_extractpath)
+            return filename[0]
+
+    def _fetchCompiler(self, extracted_tar):
+        original_path = os.getcwd()
+        with cd(self.toolchain_extractpath):
+            with cd(extracted_tar):
+                for root, dirnames, _ in os.walk('.'):
+                    for dirname in dirnames:
+                        if dirname == 'bin':
+                            with cd(original_path):
+                                return self._getCompilerFromBinaries(
+                                    os.path.join(self.toolchain_extractpath,
+                                                 extracted_tar, dirname))
+        raise ImportError('Frontend not found...')
+        return None
+
+    def _validate_compiler_model(self, model_name):
```

Comment:
hum, this is duplicated from the controller... we may need an extra helper class that will do this for both controller and factory